src/growable.c

```diff
@@ -0,0 +1,276 @@
+#include "data.table.h"
+
+#ifndef USE_GROWABLE_ALTREP
+
+SEXP growable_allocate(SEXPTYPE type, R_xlen_t size, R_xlen_t max_size) {
+  SEXP ret = PROTECT(allocVector(type, max_size));
+  SET_TRUELENGTH(ret, max_size);
+#if R_VERSION >= R_Version(3, 4, 0)
+  SET_GROWABLE_BIT(ret);
+#endif // otherwise perceived memory use will end up higher
+  SETLENGTH(ret, size);
+  UNPROTECT(1);
+  return ret;
+}
+
+R_xlen_t growable_max_size(SEXP x) {
+  return TRUELENGTH(x);
+}
+
+void growable_resize(SEXP x, R_xlen_t newsize) {
+  R_xlen_t max_size;
+  if (newsize > (max_size = growable_max_size(x))) internal_error(
+    __func__, "newsize=%g > max_size=%g",
+    (double)newsize, (double)max_size
+  );
+  SETLENGTH(x, newsize);
+}
+
+Rboolean is_growable(SEXP x) {
+  return isVector(x) && TRUELENGTH(x) >= XLENGTH(x)
+#if R_VERSION >= R_Version(3, 4, 0)
+    && IS_GROWABLE(x)
+#endif
+  ;
+}
+
+// Assuming no ALTREP columns
+SEXP make_growable(SEXP x) {
+  if (TRUELENGTH(x) < XLENGTH(x)) SET_TRUELENGTH(x, XLENGTH(x));
+  SET_GROWABLE_BIT(x);
```

Comment:
```suggestion
#if R_VERSION >= R_Version(3, 4, 0)
  SET_GROWABLE_BIT(ret);
#endif //
```

Reply:
We've got a dummy `#define` here:
https://github.com/Rdatatable/data.table/blob/9fe1b8d01701d2b23adcff3085ea0290da5df022/src/data.table.h#L17-L19

...but it may be cleaner to check the version at the point of use. Which may become moot with the move to R > 3.4.

Reply:
I see, TY!